README.md
# Old Utilities

These are old utility scripts, the oldest of which date to the late 1990s.  They were originally hosted at software.cedar-solutions.com.  

Many of these utilities are obsolete, either because of their language (i.e.  Python 2), the system they interact with (i.e. the now-defunct Gallery v2 photo gallery), or the APIs they require (i.e. deprecated Google APIs).  I am retaining them for historical reference, because even if they are obsolete they do sometimes represent a good example of how to accomplish a specific task.

See also: the [vim-syntax](https://github.com/pronovic/vim-syntax) repository, which includes Vim syntax files originally published alongside these utilities.

> _Note:_ this code is not maintained and there is no guarantee that any of it works properly any more.  Make sure you understand what a script does before you try to use it.

## Available Utilities

**batchscan**: Shell script to use SANE for scanning a batch of photographs. You need sane-utils and libtiff-tools installed.

**change-string**: Searches through a list of documents for a strings matching a Perl regular expression, and replaces the matching strings with a constant pattern. Useful when (for example) changing the copyright or author in all files in a given directory. Written in Perl 5.

**c-stats**: Acts on C or C++ source files and returns metrics related to the files. Returned metrics include overall lines, actual LOC (lines of code), number of comment-only lines and number of non-comment whitespace lines. Written in Perl 5.

**dedup.py**: Deduplicates a set of files, tossing the duplicates into a separate directory. Duplicate files are identified using an SHA256 secure hash. I have used this script to weed out duplicate photographs. Written in Python 2.

**gallery-extract**: Extracts Gallery (v1) albums into a directory and generates a printable index including captions. I used this when I want to put an album or albums onto a disc to show on a TV using someone's DVD player. It's tedious to copy the files by hand, so this script parses the Gallery data files, dumps the pictures into a directory, and renames them so the correct viewing order is obvious to the DVD player. Based on the script php-deserialize.pl by Bharat Mediratta. Written in Perl 5. _Note: In early 2015, I discovered that this script is obsolete and no longer reads the serialized file format generated by the latest version of PHP._

**gmailbackup**: Wrapper around getmail to back up gmail via IMAP. Written in Python (requires 2.5 or better). Make sure you read the notes at the top of the script before using it.

**gmailwhitelist**: Generates an email address whitelist based on gmail data. The whitelist will contain email addresses from all of your "trusted" gmail contacts - the ones that show up under "My Contacts" in the gmail user interface. Written in Python (requires 2.5 or better), and uses Google's gdata Python library for connectivity.

**kgrep**: kgrep searches through a file or files for a specified pattern and displays the target line containing the pattern as well as a certain number of lines on either side of the target line. GNU grep does provide this functionality with its -A, -B and -C switches, and similar functionality may be available on some platform-specific grep implementations. The main advantage to kgrep is that it is small, and can be easily used on any system that has Perl5 installed, as an alternative to building and installing the GNU grep binary. Written in Perl 5.

**muttsearch**: Wraps grepmail in a way that is easily called from a Mutt macro. Only works on mbox files, not with maildir. Written in Python (requires 2.1 or better).

**mutturl**: This script provides a target for urlview. It searches out commonly-installed browsers on your system and provides you a list to choose from (the list varies depending on whether `$DISPLAY` is set, i.e. on whether you are running an X session).  Written in Python (requires 2.1 or better).

**photocopy**: Makes greyscale photocopies to a printer using a SANE-enabled USB scanner. You will need to change the `$PRINT` variable to match the command-line for printing on your system. You may need to make other tweaks if you have more than one USB scanner. Your printer or print queue manager (i.e. CUPS) must support Postscript. Written in kshell.

**py-stats**: Acts on Python (*.py) source files and returns metrics related to the files. Returned metrics include overall lines, actual LOC (lines of code), number of comment-only lines and number of non-comment whitespace lines. Written in Python (requires 2.3 or better). Based in part on code by Leif K-Brooks, as documented in the code.

**rcsinfo**: Provides information about what files in a directory are checked out in RCS. The information is essentially a filtered and prettied-up version of the rlog output. Written in Perl 5.

**reorder.py**: Recurses through a directory and reorders all images by their EXIF creation date. Reordered images are written to a new target directory. This is useful if you have set of pictures taken by more than one camera, and you want to create a single cohesive image gallery. Written in Python 2.7 and requires PIL.

**secret-santa**: A "Secret Santa" exchange is one where a group of a people get together and each exchange gifts with one other person from that group. The exchange assignments are supposed to be secret, so that no one knows ahead of time who they will get their gift from. This script reads a configuration file and generates email assignments for a Secret Santa exchange. Written in Python (requires 2.5 or better). Also requires a local mail transfer agent (MTA) to send the notification emails.

**span**: Utility that "spans" (or partitions) the files in a directory into one or more sets of files, where the size of each set is as close as possible to a given capacity. When the spanning process is complete, the program writes the set definitions to disk to be used later. Four different knapsack algorithms (first-fit, best-fit, worst-fit and alternate-fit) are available to vary the method used to fit files into the sets. Written in Python (requires 2.1 or better).

**whitelist**: Processes one or more hand-maintained lists, Mutt-style addressbooks or mbox sent-mail folders and merges them into a single email whitelist. The whitelist will contain unique emails from among all chosen sources, including all To, CC and BCC emails from the messages in the sent mail folders. It is useful for mail processing using procmail, and some instructions are included. Written in Python (requires 2.4 or better) and also requires the formail utility.

**xcf2png**: Combination of a shell script and a Gimp script-fu script that converts Gimp XCF files (i.e. native Gimp files) to PNG format. Place the script-fu script in your `~/gimp/scripts` directory before running the shell script. Based in part on a different script with the same name by Rik Hemsley, as documented in the code.


secret-santa/Changelog
Version 1.5    01 Dec 2010

	* First time looking at this in a while.
	* Run pylint on code and fix warnings.
	* Standardize the headers, clean up comments, etc.

Version 1.3    07 Dec 2005

	* Add -c,--autoconflict option to script (closes: #14).
	* Modify secret-santa.1 manpage to document new option.
	* Modify secret-santa.1 to remove references to Cedar Solutions, Inc.
	* Modify secret-santa.conf.5 to remove references to Cedar Solutions, Inc.
	* Remove README.Developers from Python source distribution.
	* Clean up developer-private makefile.
	* Convert Changelog to match Vim syntax format.

Version 1.2    08 Sep 2003

	* Converted file headers to new Cedar Solutions standard.
	* Added email address to all files where author is listed.
	* Removed all references to Cedar Solutions, Inc.
	* Added encoding to top of all Python files.
	* Did cleanup of some of the private maintainer scripts.

Version 1.1    10 Dec 2002

	* Documentation now mentions requirement on an MTA such as sendmail or exim.
	* Added file README.Developers

Version 1.0    09 Dec 2002

	* Initial "public" release.

-----------------------------------------------------------------------------
vim: set ft=changelog noexpandtab:


dedup.py
#!/usr/bin/python
# -*- coding: iso-8859-1 -*-
# vim: set ft=python ts=3 sw=3 expandtab:
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
#
#              C E D A R
#          S O L U T I O N S       "Software done right."
#           S O F T W A R E
#
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
#
# Copyright (c) 2015 Kenneth J. Pronovici.
# All rights reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License,
# Version 2, as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# Copies of the GNU General Public License are available from
# the Free Software Foundation website, http://www.gnu.org/.
#
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
#
# Authors  : Kenneth J. Pronovici <pronovic@ieee.org>
# Language : Python (>= 2.7)
# Revision : $Id: dedup.py 1378 2015-02-21 21:28:16Z pronovic $
# Purpose  : De-duplicate a set of files in a directory
#
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

########################################################################
# Imported modules
########################################################################

import hashlib
import glob
import os.path
import shutil
from optparse import OptionParser


########################################################################
# Utility code
########################################################################

# Memory-efficient way to get crypographic hash for a file
# @param path  Path to a file on disk
# See: http://stackoverflow.com/questions/3431825/generating-a-md5-checksum-of-a-file
def cryptohash(path):
   blocksize = 65536
   hasher = hashlib.sha256()
   with open(path, 'rb') as fp:
      buf = fp.read(blocksize)
      while len(buf) > 0:
         hasher.update(buf)
         buf = fp.read(blocksize)
      return hasher.hexdigest()

# Parses command-line arguments.
# This function will not return if command-line arguments are invalid.
# @return Tuple of (duplicatesDir, source-glob)
def handleCommandLine():
   usage = "%prog [--help] duplicatesDir source-glob"
   optionParser = OptionParser(usage=usage)
   (options, args) = optionParser.parse_args()
   if len(args) != 2:
      optionParser.parse_args(["--help"])  # convenient way to get the error message
   return (args[0], args[1])


########################################################################
# Main routine
########################################################################

# Main routine for program
def main():
   (duplicatesDir, sourceGlob) = handleCommandLine()

   print ""
   print "De-duplicating %s" % sourceGlob
   print "Duplicates written to %s" % duplicatesDir

   print ""
   print "Generating crypto hashes for files..."
   hashes = { }
   for path in glob.iglob(sourceGlob):
      hashes[path] = cryptohash(path)

   print ""
   print "Checking for duplicates..."
   for i in glob.glob(sourceGlob):
      if os.path.exists(i):
         print "   Checking %s, hash %s" % (os.path.basename(i), hashes[i])
         for j in glob.glob(sourceGlob):
            if os.path.exists(j):
               if i != j: 
                  if hashes[i] == hashes[j]:
                     print "      Found duplicate %s" % os.path.basename(j)
                     shutil.move(j, duplicatesDir)
   

########################################################################
# Module entry point
########################################################################

# Run the main routine if the module is executed rather than sourced
if __name__ == '__main__':
   main()



reorder.py
#!/usr/bin/python
# -*- coding: iso-8859-1 -*-
# vim: set ft=python ts=3 sw=3 expandtab:
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
#
#              C E D A R
#          S O L U T I O N S       "Software done right."
#           S O F T W A R E
#
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
#
# Copyright (c) 2010-2011 Kenneth J. Pronovici.
# All rights reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License,
# Version 2, as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# Copies of the GNU General Public License are available from
# the Free Software Foundation website, http://www.gnu.org/.
#
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
#
# Authors  : Kenneth J. Pronovici <pronovic@ieee.org>
# Language : Python (>= 2.7), plus PIL (Python Imaging Library)
# Revision : $Id: reorder.py 1336 2013-07-27 19:16:21Z pronovic $
# Purpose  : Reorder images by their EXIF creation date.
#
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

########################################################################
# Notes
########################################################################

"""
This program reorders the images in a directory by their EXIF creation date.
This is useful if you are trying to cull through a set of pictures taken by
more than one camera and turn them into one cohesive image gallery.
@author: Kenneth J. Pronovici <pronovic@ieee.org>
"""


########################################################################
# Imported modules
########################################################################

# System modules
import sys
import os
import re
import math
import shutil
from datetime import datetime
from datetime import timedelta

# PIL modules (Python Imaging Library)
import Image   
from ExifTags import TAGS


#######################################################################
# Functions
#######################################################################

def usage():
   """Prints out program usage information."""
   print ""
   print "Usage: %s source-dir target-dir [offsets]" % os.path.basename(sys.argv[0])
   print ""
   print "Finds all image files in source-dir and reorders them into"
   print "target-dir based on EXIF creation date.  The target directory"
   print "will be created if it does not already exist."
   print ""
   print "You may optionally provide offsets by camera model, in case"
   print "the clocks on your cameras are not in sync.  Use the format"
   print "\"model=+/-hours:minutes\", for instance \"PowerShot A70=+06:55\"."
   print "The configured number of hours and minutes will be added to or"
   print "subtracted from the actual time on the camera."

def findImages(sourcedir, offsets):
   """
   Recurses through a directory, building a list of image files in it.
   @param sourcedir: Source directory to recurse through
   @param offsets: Configured offsets as from parseOffsets()
   @returns: List of (filename, EXIF creation date) tuples.
   """
   images = []
   def visit(items, dir, entries):
      for entry in entries[:]:
         filename = os.path.join(dir, entry)
         if os.path.isfile(filename):
            exifCreationDate = getExifCreationDate(filename, offsets)
            if exifCreationDate is not None:
               images.append((filename, exifCreationDate))
   os.path.walk(sourcedir, visit, images)
   return images

def copyImages(images, targetdir):
   """
   Copies images into the target directory, ordering them by EXIF creation date.
   @param images: List of image tuples, as from findImages()
   @param targetdir: Target directory to write images into
   """
   prefix = "image"
   if not os.path.exists(targetdir):
      os.makedirs(targetdir)
   digits = digitsRequired(len(images))
   images.sort(key=lambda x: x[1])
   index = 0
   for image in images:
      index += 1
      source = image[0]
      prepend = prefix + '{0:0{digits}}__'.format(index, digits=digits)
      target = os.path.join(targetdir, prepend + os.path.basename(source))
      shutil.copyfile(source, target)
#      print "%s\t%s\t%s" % (image[0], image[1], prepend + os.path.basename(source))

def getExifCreationDate(filename, offsets):
   """
   Get the EXIF creation date for a file.
   If the file is not an image, then we'll just return None.
   @param filename: Absolute path to a file
   @param offsets: Configured offsets as from parseOffsets()
   @return: EXIF creation date, or None if there is no such information.
   """
   dateTime = None
   cameraModel = None
   try:
      i = Image.open(filename)
      info = i._getexif()
      keymap = { }
      if info is not None:
         for tag, value in info.items():
            decoded = TAGS.get(tag, tag)
            keymap[decoded] = value
         # Some cameras have DateTimeOrignal, and some only have DateTime.  On
         # the ones that have DateTimeOriginal, it looks like DateTime might be
         # different (newer), possibly because the file was modified on the
         # camera or something.  We definitely want to use DateTimeOriginal if
         # we have it.
         if keymap.has_key("DateTimeOriginal"):
            dateTime = keymap["DateTimeOriginal"]
         elif keymap.has_key("DateTime"):
            dateTime = keymap["DateTime"]
         if keymap.has_key("Model"):
            cameraModel = keymap["Model"]
   except IOError: pass
   if dateTime is None:
      sys.stderr.write("Warning, file has no date/time: %s\n" % filename)
      return None
   else:
      exifCreationTime = datetime.strptime(dateTime, "%Y:%m:%d %H:%M:%S")
      if cameraModel in offsets:
         exifCreationTime += offsets[cameraModel]
      return exifCreationTime

def digitsRequired(value):
   """Returns the number of digits required to represent a number."""
   return int(math.ceil(math.log(value+1, 10)))

def parseOffsets(args):
   """
   Parse offsets from the command-line.

   Offsets are int the form "model=[+-]hours:minutes".  For instance, use
   "Canon PowerShot A70=+06:55" to add 6 hours 55 minutes to the times in
   images shot by model "Canon PowerShot A70".  Of course, this falls apart
   if you have two separate cameras that have the same model.
   
   @param args: Command-line arguments to be parsed.

   @return: Map of model to timedelta that should be added to the actual time.
   """
   offsets = { }
   pattern = re.compile(r"(^)(.*)(=)([+-])([0-9][0-9])(:)([0-9][0-9])($)")
   for arg in args:
      result = pattern.match(arg)
      if result is None:
         raise Exception("Invalid offset configuration")
      cameraModel = result.group(2)
      plusOrMinus = result.group(4)
      hours = int(result.group(5))
      minutes = int(result.group(7))
      delta = timedelta(hours=hours, minutes=minutes)
      if plusOrMinus == "-":
         delta *= -1
      offsets[cameraModel] = delta
   return offsets


##################
# main() function
##################

def main():

   """Main routine for program."""

   offsets = { }

   try:
      sourcedir = sys.argv[1]
      targetdir = sys.argv[2]
      offsets = parseOffsets(sys.argv[3:])
   except Exception, e:
      usage()
      sys.exit(1)

   print "Building list of image files in: %s" % sourcedir
   images = findImages(sourcedir, offsets)
   if len(images) == 0:
      print "Completed -- no images to process"
   else:
      print "Copying %d images to target: %s" % (len(images), targetdir)
      sets = copyImages(images, targetdir)
      print "Completed -- finished copying all images"
      

########################################################################
# Module entry point
########################################################################

# Run the main routine if the module is executed rather than sourced
if __name__ == '__main__':
   main()



xcf2png.scm
; vim: set ft=scheme:
; # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
; #
; #              C E D A R
; #          S O L U T I O N S       "Software done right."
; #           S O F T W A R E
; #
; # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
; #
; # Copyright (c) 2002-2003 Kenneth J. Pronovici.
; # All rights reserved.
; #
; # Portions of this code are based on a similar utility that is
; # (C) Copyright 1999 Rik Hemsley <rik@kde.org> and believed to
; # be released under the GPL.
; #
; # This program is free software; you can redistribute it and/or
; # modify it under the terms of the GNU General Public License,
; # Version 2, as published by the Free Software Foundation.
; #
; # This program is distributed in the hope that it will be useful,
; # but WITHOUT ANY WARRANTY; without even the implied warranty of
; # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
; #
; # Copies of the GNU General Public License are available from
; # the Free Software Foundation website, http://www.gnu.org/.
; #
; # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
; #
; # Author   : Kenneth J. Pronovici <pronovic@ieee.org>
; # Language : Gimp script-fu language (nomimally Scheme)
; # Project  : Gimp graphics utilities
; # Package  : xcf2png
; # Revision : $Id: xcf2png.scm,v 1.5 2003/09/08 20:39:40 pronovic Exp $
; # Purpose  : script-fu script to convert from XCF to PNG
; #
; # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

; ######################################################################
; # Notes
; ######################################################################
;
; This file is meant to be used with the xcf2png kshell script.  Place
; this file in ~/gimp-1.2/scripts (or the appropriate directory for your
; version of Gimp) before running the shell script.
;
; Call (xcf2png "file.xcf" "file.png" "file.xcf.temp") to convert
; file.xcf to PNG format using the highest possible compression.
; Remember to remove file.xcf.temp when you're done.
;
; It should be possible to make the Gimp flatten an XCF image and then
; save it as a PNG file.  However, I can't seem to make it do that in
; just a single step.  Instead, I first save a temporary XCF file that
; is flattened, and then convert that flattened file to PNG format.
;
; I am not much of a Scheme coder.  I guess the code below looks like
; I'm trying to program in C, instead.  Sorry...
;

; ############################
; # _xcf2png_flatten function
; ############################
; Takes an XCF file, flattens it, and saves it to a temporary file

(define ( _xcf2png_flatten xcffile tempfile)
  (let* ( 
          ( image ( car ( gimp-xcf-load 0 xcffile tempfile ) ) )
          ( drawable ( car ( gimp-image-active-drawable image ) ) ) 
        )
    (if ( > ( car ( gimp-image-get-layers image ) ) 1 )
      (gimp-image-flatten image )
    )
    ( gimp-xcf-save 1 image drawable tempfile tempfile ) 
  ) 
)

( script-fu-register "_xcf2png_flatten" 
                     "<Toolbox>/_xcf2png_flatten" 
                     "" "" "" "" "" )


; ############################
; # _xcf2png_flatten function
; ############################
; Converts a flattened XCF file to PNG format.

(define ( _xcf2png_convert xcffile pngfile)
   (let* ( 
           ( image ( car ( gimp-xcf-load 0 xcffile pngfile ) ) )
           ( drawable ( car ( gimp-image-active-drawable image ) ) )
         )
      ( file-png-save 1 image drawable pngfile pngfile 0 9 0 0 0 0 0 )
   ) 
)

( script-fu-register "_xcf2png_convert" 
                     "<Toolbox>/_xcf2png_convert" 
                     "" "" "" "" "" )

; ###################
; # xcf2png function
; ###################
; Converts an XCF file to PNG format

( define ( xcf2png xcffile pngfile tempfile)
  ( _xcf2png_flatten xcffile tempfile)
  ( _xcf2png_convert tempfile pngfile)
  ( gimp-quit 0 ) 
)

( script-fu-register "xcf2png" 
                     "<Toolbox>/xcf2png" 
                     "" "" "" "" "" )

